jetty-ee10/jetty-ee10-annotations/src/main/java/org/eclipse/jetty/ee10/annotations/AnnotationParser.java

```diff
@@ -717,92 +559,66 @@ public void parse(final Set<? extends Handler> handlers, Resource r) throws Exce
         if (r == null)
             return;
 
-        if (r.exists() && r.isDirectory())
+        if (!r.exists())
+            return;
+
+        if (FileID.isJavaArchive(r.getPath()))
         {
-            parseDir(handlers, r);
+            parseJar(handlers, r);
             return;
         }
 
-        String fullname = r.toString();
-        if (fullname.endsWith(".jar"))
+        if (r.isDirectory())
         {
-            parseJar(handlers, r);
+            parseDir(handlers, r, r.getPath());
             return;
         }
 
-        if (fullname.endsWith(".class"))
+        if (FileID.isClassFile(r.getPath()))
         {
-            try (InputStream is = Files.newInputStream(r.getPath()))
-            {
-                scanClass(handlers, null, is);
-                return;
-            }
+            parseClass(handlers, null, r.getPath());
         }
 
         if (LOG.isDebugEnabled())
-            LOG.warn("Resource not scannable for classes: {}", r);
+            LOG.warn("Resource not able to be scanned for classes: {}", r);
     }
 
     /**
      * Parse all classes in a directory
      *
      * @param handlers the set of handlers to look for classes in
-     * @param root the resource directory to look for classes
+     * @param baseResource the resource representing the baseResource being scanned (jar, dir, etc)
+     * @param dir the directory root to start the scanning of classes for
      * @throws Exception if unable to parse
      */
-    protected void parseDir(Set<? extends Handler> handlers, Resource root) throws Exception
+    private void parseDir(Set<? extends Handler> handlers, Resource baseResource, Path dir) throws Exception
```

Comment:
Why is there both `baseResource` and `Path`. Surely they are one and the same thing?

Reply:
This can be different in the case of a JAR.
The `baseResource`, is the JAR itself (the File)
The `dir` is the Directory (as a Path) being scanned, which is the root of the archive.

Ideally, I want to eliminate the tracking of the Resource, as it's a needless bit of trivia in the AnnotationParser IMO.

Reply:
Turns out there is 1 reason for `Resource` being tracked as `baseResource`, and that's a single code block down in `org.eclipse.jetty.ee10.webapp.MetaData` in it's `addDiscoveredAnnotation(DiscoveredAnnotation annotation)` method.

https://github.com/eclipse/jetty.project/blob/jetty-12.0.x/jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/MetaData.java#L360-L390

Oddly, if I intentionally break that code block with a null resource and run the build/test, the only tests that fail are 2 test methods in `org.eclipse.jetty.ee10.webapp.TestMetaData`.
(testAddDiscoveredAnnotation and testResolve).   No other tests fail?  Does this mean this requirement for tracking a Resource only exists to satisfy a testcase?

Something to investigate later, for now I'm leaving the tracking of Resource in place.

Reply:
No, the requirement for tracking the resource is used eg in Quickstart, where we set an attribute with the source (descriptor/annotation/programmatic) of each element that is generated in the quickstart-web.xml. Please keep it.

Reply:
Also, I still don't see why you need to pass in the `Path`, why can't you just ask the `baseResource` for it's `Path`? You're only ever passing in the root level of the jarfile, so it's not like this method is scanning some subdirectory of the `baseResource`, it's scanning the `baseResource` itself.

Reply:
> No, the requirement for tracking the resource is used eg in Quickstart, where we set an attribute with the source (descriptor/annotation/programmatic) of each element that is generated in the quickstart-web.xml. Please keep it.

Can you point out where it's used?
I cannot find where the resource is ever used in quickstart, not even from a `.toString()`

As there is currently a difference between `baseResource` and `dir`, they point to different things when it's a JAR. (they point to the same thing if the resource is a directory. eg: `file://path/to/dir/`)

`baseResource` is the JAR itself, the File, aka `file://path/to/foo.jar`
`dir` is a directory, pointing to the root contents of the JAR, as in `jar:file://path/to/foo.jar!/`

This is an important distinction in Resource.
One is a File, no mount needed, it's treated like a file on disk, nothing more.
The other is a directory, with a mount being managed.

I need to know how quickstart uses this resource, to know if I should be passing in the `file://path/to/foo.jar` or `jar:file://path/to/foo.jar!/`.

Reply:
Checking `jetty-10.0.x` and it's `test-quickstart` i see output like this ...

``` xml
  <context-param>
    <param-name>org.eclipse.jetty.resources</param-name>
    <param-value><![CDATA[
    "jar:${WAR.uri}/WEB-INF/lib/demo-web-fragment-10.0.12-SNAPSHOT.jar!/META-INF/resources"]]></param-value>
  </context-param>
```

Is this what you are referring to?

Reply:
Looks like Quickstart generates optional the `origin="..."` attributes, that has references to resource.
These are informational, and nothing in our code uses those attributes.

Good, that means `parseDir()` can use just the Resource, even if it's `file://` or `jar:file://` without breaking quickstart.  

I'll make this change.

Reply:
Change made in commit a5eeac1c786e1bb192da184cb9b4986e49eb730c